File: Ai algorithm implementation/Stochastic_Hill_Climbing_(N-Queens).cpp
#include <bits/stdc++.h>
using namespace std;

int N = 8;

// Calculate attacking pairs
int heuristic(vector<int> &state) {
    int h = 0;
    for(int i=0;i<N;i++)
        for(int j=i+1;j<N;j++)
            if(state[i]==state[j] || abs(state[i]-state[j])==abs(i-j)) h++;
    return h;
}

vector<int> stochasticHC(vector<int> state) {
    srand(time(0));
    while(true) {
        vector<pair<int,int>> improvingMoves;

        for(int col=0; col<N; col++) {
            int originalRow = state[col];
            for(int row=0; row<N; row++) {
                if(row == originalRow) continue;
                state[col] = row;
                int h = heuristic(state);
                if(h < heuristic(state)) improvingMoves.push_back({col,row});
            }
            state[col] = originalRow;
        }

        if(improvingMoves.empty()) break;

        // pick random improving move
        auto move = improvingMoves[rand()%improvingMoves.size()];
        state[move.first] = move.second;
    }
    return state;
}

int main() {
    vector<int> state(N);
    srand(time(0));
    for(int i=0;i<N;i++) state[i] = rand()%N;

    vector<int> solution = stochasticHC(state);
    cout << "Stochastic Hill Climbing Solution (Heuristic=" << heuristic(solution) << "):\n";
    for(int i=0;i<N;i++) cout << solution[i] << " ";
    cout << "\n";
}

File: Ai algorithm implementation/Stochastic_Hill_Climbing_(N-Queens).py
import random

N = 8

def heuristic(state):
    h = 0
    for i in range(N):
        for j in range(i + 1, N):
            if state[i] == state[j] or abs(state[i] - state[j]) == abs(i - j):
                h += 1
    return h

def stochastic_hc(state):
    while True:
        current_h = heuristic(state)
        improving_moves = []
        
        for col in range(N):
            original_row = state[col]
            for row in range(N):
                if row == original_row:
                    continue
                state[col] = row
                h = heuristic(state)
                if h < current_h:
                    improving_moves.append((col, row))
            state[col] = original_row
        
        if not improving_moves:
            break
        
        move = random.choice(improving_moves)
        state[move[0]] = move[1]
    
    return state

random.seed()
state = [random.randint(0, N - 1) for _ in range(N)]

solution = stochastic_hc(state)
print(f"Stochastic Hill Climbing Solution (Heuristic={heuristic(solution)}):")
print(solution)
